Fix latex issues via value added scripts

michael-okeefe opened the pull request:
Pull request overview
---------------------
 - Fixes #8026 

### Purpose and Approach

Fixes Misc. LaTeX Errors Not Caught in Normal Workflow by adding a python-based LaTeX log parsing capability in Python that summarizes the relevant issues and warnings in the LaTeX log files by file and line number and also adds a JSON summary of the specific issues that can be further processed and/or displayed by the CI. This pull request also then fixes the majority of these issues across all the docs. These issues include package-based issues and warnings (for e.g., issues reported by \siunits) as well as issues and warnings from both the TeX and LaTeX processing pipeline.

The only new code added is the `doc/tools/parse_latex_log.py` file. That file uses no external dependencies beyond the python standard library. Fixes to the docs were found and fixed using the errors and warnings reported by the LaTeX log parser. The hope is that this script will make LaTeX errors and warnings more visible and actionable for future EnergyPlus documentation authors.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [X] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

michael-okeefe committed:
Add python script to parse the LaTeX logs

michael-okeefe committed:
Add latex log parser updates to write error file, etc

michael-okeefe committed:
Update latex log parser per comments from Edwin

michael-okeefe committed:
Correct incorrect parsing of the current file when file path wraps over more than one line

michael-okeefe committed:
remove error files before running and potentially writing

michael-okeefe committed:
handle blank line after start of a tex issue

michael-okeefe committed:
Parse general LaTeX warnings.

michael-okeefe committed:
Switch to a class-based parser.

michael-okeefe committed:
Modify script to only write JSON file and to stream issues to stdout

The script now only will write a JSON file if issues are found. If no issues are found, no file is written. "Text" based log messages are streamed to stdout, one issue per line. If issues are found, a non-zero exit code is used for helping ctest detect issues found.

michael-okeefe committed:
Add CTest test for docs that uses the logging script.

michael-okeefe committed:
Remove extraneous newlines at beginning and end of error messages.

michael-okeefe committed:
Correct TeX file parsing and error/warning designations.

michael-okeefe committed:
Add error handling and explicit encoding on windows

michael-okeefe committed:
Add fixes found from going through the LaTeX logs using the new system.

michael-okeefe committed:
Merge branch 'develop' into fix-latex-issues-via-value-added-scripts

nealkruis committed:
Set DOCS_TESTING on linux CI where DOCS are built.

nealkruis committed:
Merge remote-tracking branch 'nrel/develop' into fix-latex-issues-via-value-added-scripts

michael-okeefe committed:
Update to not use f-strings (for compatibility with Python 3.5)

michael-okeefe committed:
Merge branch 'develop' into fix-latex-issues-via-value-added-scripts

michael-okeefe committed:
Remove pathlib for CI, remove unused pdb, address flake8 issues

michael-okeefe committed:
Upgrade python dependency to 3.6 and call it via cmake's find-package

Note: a second clause is also added to docs/CMakeLists.txt should a user be doing a "docs-only" build.

michael-okeefe committed:
Update latex parser to output full relative paths to the issue tex file when run

michael-okeefe committed:
Merge branch 'develop' into fix-latex-issues-via-value-added-scripts

michael-okeefe committed:
Remove printing of extraneous information

michael-okeefe committed:
Merge branch 'develop' into fix-latex-issues-via-value-added-scripts

mitchute commented on doc/acknowledgments/src/notices-and-acknowledgments/other-software-included-or-referenced.tex:
```diff
@@ -128,7 +128,7 @@ \subsection{MODELISAR Notice}\label{modelisar-notice}
 This file is licensed by the copyright holders under the BSD License\footnote{\url{http://www.opensource.org/licenses/bsd-license.html\textgreater}}:
 
-\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}
+\begin{center}\rule{0.5\linewidth}{0.4pt}\end{center}
```
What's the purpose of this change?

michael-okeefe replied:
This is a fix for a warning generated by LaTeX. `\linethickness` is not a valid option for the second argument to `\rule` but the LaTeX compiler typically muddles through. Just fixing in the name of removing warnings from the LaTeX compiler.

mitchute commented on doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex:
```diff
@@ -111,10 +111,10 @@ \subsection{Sky Radiation Modeling}\label{sky-radiation-modeling}
 \noindent Example:
 \newline
 
-Clear sky ($N=0$), $T_{db} = 20 + 273.15 = \SI{293.15}{\kelivin}$, $T_{dp} = 10 + 273.15 = \SI{283.15}{\kelvin}$
```
Ah ```kelivin```, almost Kevin's favorite number. https://www.urbandictionary.com/define.php?term=Keleven

mitchute commented on doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/climate-calculations.tex:
```diff
@@ -111,10 +111,10 @@ \subsection{Sky Radiation Modeling}\label{sky-radiation-modeling}
 \newline
 
-$\epsilon_{sky} = \left(0.787 + 0.764  \ln\left(283.15/273\right)\right)  \left(1 + 0.0224  N - 0.0035  N^2 + 0.00028  N^3) = 0.815$
+$\epsilon_{sky} = \left(0.787 + 0.764  \ln\left(283.15/273\right)\right)  \left(1 + 0.0224  N - 0.0035  N^2 + 0.00028  N^3 \right) = 0.815$
```
Nice to see these balanced parentheses being found.

mitchute commented on doc/engineering-reference/src/daylighting-and-window-calculations/time-step-daylighting-calculation.tex:
```diff
@@ -196,23 +196,23 @@ \subsection{Interior Illuminance}\label{interior-illuminance}
 
 \begin{equation}
-  {I_{tot}}({i_L}) = \sum\limits_{\scriptstyle{\rm{windows}}\atop\scriptstyle{\rm{in~zone}}} {{I_{win}}({i_s},{i_L})}
+  {I_{tot}}({i_L}) = \sum\limits_{\genfrac{}{}{0pt}{2}{\rm{windows}}{\rm{in~zone}}} {{I_{win}}({i_s},{i_L})}
```
Much better.

mitchute commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-001/coils.tex:
```diff
@@ -1611,38 +1611,38 @@ \subsubsection{Special Calculations for Coil:Cooling:DX:TwoStageWithHumidityCont
 1. No latent control
 
-   Perform normal operation with SHR_{Normal}
+   Perform normal operation with SHR\(_{Normal}\)
```
The output of these changes look identical on my machine. What's the purpose?

michael-okeefe replied:
Another warning and, again, the LaTeX compiler is able to guess in this instance and muddle through. What's wrong is the `_{blah}` stuff is not valid outside of math mode. The `\(` and `\)` puts us in math mode (similar to `$`...`$` but allows checking for balancing).

mitchute commented on doc/input-output-reference/src/hvac-template-objects/group-hvac-templates.tex:
```diff
@@ -7557,7 +7557,7 @@ \subsubsection{Inputs}\label{inputs-25-000}
 
 The name of this object.
 
-\paragraph{Field: Chiller Object Type}\label{field-chiller-object-type}
```
Excellent. No need for unnecessary label references.

mitchute commented on doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex:
```diff
@@ -863,7 +863,7 @@ \subsubsection{Inputs}\label{inputs-8-011}
 \textbf{Idso} -- the clear sky emissivity are calculated using Idso model.
 
-\textbf{BerdahlMartin} -- the clear sky emissivity are calculated using Berdahl & Martin model.
+\textbf{BerdahlMartin} -- the clear sky emissivity are calculated using Berdahl \& Martin model.
```
Good catch.

mitchute commented on doc/input-output-reference/src/overview/group-simulation-parameters.tex:
```diff
@@ -963,7 +963,7 @@ \subsection{PerformancePrecisionTradeoffs}\label{performanceprecisiontradeoffs}
 
 \begin{equation}
-{\left {PLR}\right} = \frac{SystemLoad-OutputOff}{OutputFull - OutputOff}
+PLR = \frac{SystemLoad-OutputOff}{OutputFull - OutputOff}
```
The output of this change looks identical. Regardless, this is a good cleanup.

michael-okeefe replied:
Yeah, this is yet another instance where the LaTeX compiler issues a warning because the syntax is incorrect. If we really wanted to wrap PLR in "{}", we need to wrap it in `\left\{` and `\right\}`. However, I was assuming the wrapping of {} wasn't intentional. Cleaning up LaTeX issues where the compiler muddles through (so output doesn't change) but where the syntax as written is incorrect does really help with importing the LaTeX for the web-based documentation. Thanks!

mitchute reviewed (commented):
This is a nice set of changes that should help enable us to clean up documentation issues. 👍  👍

nealkruis committed:
Automatically set DOCS_TESTING from BUILD_DOCS and BUILD_TESTING.

nealkruis committed:
Merge remote-tracking branch 'nrel/develop' into fix-latex-issues-via-value-added-scripts

Myoldmopar commented:
This looks great!  CI caught the warnings and parsed them out nicely.  However, we need to remove the other warnings that were already present before the CI improvements.  If you can clean those up, I think we'll be good to go.

Myoldmopar commented:
And to be clear, there are a couple options:
 - Make a commit that cleans out just the extra `LATEXWARNING::` stuff that is showing up, but leaves the LaTeX "mistakes" in place so that we demonstrate perfect CI behavior.  Then follow it up with a commit that gets the LaTeX code looking all clean.  Or...
 - Just do both of those at once and get a clean result.

I don't care either way.  Once both of those steps are done, however, mark this as ready for review and we'll get it in.

Thanks @nealkruis @michael-okeefe

nealkruis commented:
We can either pipe stdout into a txt file, or have an option to silence the python script output.

If I'm understanding things correctly, that would mean that we would see the three tests fail, but there wouldn't be any output below them. This would be followed by the new failure messages from parsing the *_errors.json files.

We could change it so that executing the python script is just another custom target or custom command in CMake instead of a CTest. That way you wouldn't see the extra failed tests.

Either way, once we get that sorted out, we'll clean up the actual LaTeX warnings/errors so this can go in cleanly.

Myoldmopar commented:
> We could change it so that executing the python script is just another custom target or custom command in CMake instead of a CTest. That way you wouldn't see the extra failed tests.

Yep, this would be ideal.  Please make it so, but leave the LaTeX warnings in place, and we'll make sure it all looks good, then do one clean commit and merge.

Thanks!

nealkruis committed:
Make LaTeX error logs as a post-build step instead of CTest.

nealkruis commented:
Results are in! It definitely prevents other tests from being run successfully if python sends a bad return code.

@michael-okeefe let's set it up to always return zero so it doesn't prevent other tests from running.

@Myoldmopar since these aren't really Test Results, would it make sense to have the CI put them somewhere else (e.g., in "Build Messages" or in a separate section just for documentation)?

michael-okeefe committed:
Prevent LaTeX log parser from ever returning a non-zero exit code to prevent stopping ctest when run

michael-okeefe commented:
I just modified the script so that it never returns a non-zero exit code and pushed that up (it was already on-par with the latest from develop). Shall I also start cleaning up the remaining LaTeX issues that were identified?

Myoldmopar commented:
This looks fantastic.  Once you address those LaTeX issues and we get a clean build we'll merge this in.  Thanks for the refinements here, @nealkruis @michael-okeefe !

Myoldmopar commented:
Oh, and about your comment on the build results/test results.  Unfortunately, since this is run as a post-processing ctest, I think it's too late for CI to grab these as "build results".  But they do show up fine as parsed test results.

michael-okeefe commented:
> This looks fantastic. Once you address those LaTeX issues and we get a clean build we'll merge this in. Thanks for the refinements here, @nealkruis @michael-okeefe !

Sounds great. I just started on addressing the last of the LaTeX issues. Thanks!

Myoldmopar commented on CMakeLists.txt:
```diff
@@ -103,8 +103,8 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
-# we are making *a Python 3 Interpreter* a required dependency, so find it here
-find_package(PythonInterp 3 REQUIRED)
+# we are making *a Python 3.6 Interpreter* a required dependency, so find it here
+find_package(PythonInterp 3.6 REQUIRED)
```
Sometime in the future we need to move away from this PythonInterp as it was [deprecated](https://cmake.org/cmake/help/v3.12/module/FindPythonInterp.html) in 2018 in favor of `find_package(Python3)`, but first we need to move our minimum cmake up.

Myoldmopar commented on CMakeLists.txt:
```diff
@@ -337,6 +337,7 @@ if( BUILD_PACKAGE )
 endif()
 
 if (BUILD_DOCS)
+  set(DOCS_TESTING ${BUILD_TESTING} CACHE BOOL "" FORCE)
```
OK, so as soon as you enable both BUILD_DOCS and BUILD_TESTING, this will be forced to ON.  No problem.  Since this is automatically triggered to ON/OFF, I wonder if we should either mark it advanced, or just take it out of the cache entirely and treat it like a local cmake variable.

Myoldmopar commented on doc/CMakeLists.txt:
```diff
@@ -4,6 +4,13 @@ cmake_minimum_required(VERSION 3.5.1)
+if(NOT EXISTS ${PYTHON_EXECUTABLE})
+  # we are making *a Python 3.6 Interpreter* a required dependency, so find it here
+  find_package(PythonInterp 3.6 REQUIRED)
+endif()
```
It would be cool to just reuse the previously found Python, but I know that may be an annoying thing to manage, so it's OK.

Myoldmopar commented on doc/cmake/doc-targets.cmake:
```diff
@@ -11,5 +11,14 @@ macro( CREATE_DOC_TARGET SOURCE_FILENAME OUTPUT_FILENAME )
+      POST_BUILD
+      COMMAND ${PYTHON_EXECUTABLE} "${PROJECT_SOURCE_DIR}/tools/parse_latex_log.py" "${PROJECT_SOURCE_DIR}/${SOURCE_FILENAME}/${SOURCE_FILENAME}.log" "${PROJECT_SOURCE_DIR}/${SOURCE_FILENAME}" "${PROJECT_BINARY_DIR}/${OUTPUT_FILENAME}_errors.json"
+      )
+  endif()
```
OK, so if this is just a POST_BUILD on the PDF building portion, then the file should be available once "make" is complete.  In this case, I think I actually could have CI find the error files early enough to catch these as build errors, not test failures.  Let me know if I am correct in the build step and I can make an alteration to CI.

Myoldmopar commented on doc/tools/parse_latex_log.py:
```diff
@@ -0,0 +1,552 @@
+                'plant-application-guide',
+                'tips-and-tricks-using-energyplus',
+                'using-energyplus-for-compliance',
+                ]
```
Is this list really there just for testing?  Or is this something we have to maintain?

michael-okeefe replied:
Yeah, I was debating about this and ended up leaving it in as I found it useful but you are right -- it could become a maintenance issue if anyone else were to depend on it. On a second look, I'd vote to just remove it.

Myoldmopar replied:
:+1:

michael-okeefe replied:
I've fixed this and the remaining LaTeX issues that were flagged. Just pushed that up.

I also added a check in the python script to delete the target JSON file if it exists before writing as a persistent JSON error file would be confusing for those working off of their local laptops and actively fixing issues.

Myoldmopar reviewed (approved):
This looks pretty great as-is, though a couple thoughts came to mind that could be considered:
 - Do we want to make DOCS_TESTING just an internal variable, nothing exposed to the cache?
 - If the doc test is run as a post build step, I would be happy to tell CI to search for the `errors.json` files after the build rather than after the test.

michael-okeefe committed:
Remove cross-document links.

It appears there was an attempt to hyperref these from Engineering Reference to Input/Output Reference. Otherwise, there is not other label target for those. I did a case insensitive search and also searched for any label{...} with the first part of the hyperref tags.

michael-okeefe committed:
Fix issue with dropping more than two header levels.

Section-level was skipped so adjusted all heading levels accordingly.

michael-okeefe committed:
Remove hard-coded docs tags from parse_latex_log.py

Remove special case for parse_latex_log.py. Best to just use the standard build tooling via CMake.

michael-okeefe committed:
Correct multiple header level jump

The header-level jumps from Chapter to subsection and skips section level. This edit corrects that.

michael-okeefe committed:
Remove multiply-defined labels

michael-okeefe committed:
Change link from Table:OneIndependentVariable to Table:Lookup.

michael-okeefe committed:
Fix flake8 linter warnings.

michael-okeefe committed:
Prevent accidental persistence of json error file

Since the presence of the JSON file indicates that issues are found, this prevents someone from running, finding and fixing errors, and then re-running (all fixed) but the json file is still there so they think the errors haven't been fixed.

Myoldmopar commented:
This looks great!  I'm not sure why Mac hasn't reported yet, but Mac wasn't involved, so I'm OK with not waiting on Mac for this.  The only comment I had unresolved was whether we should just eliminate the DOCS_TESTING cache variable since it is controlled via the two cache variables BUILD_DOCS and BUILD_TESTING.  This doesn't have to be held up for that, but I'd like to know what you think of that change at least.

nealkruis commented:
We'll still need DOCS_TESTING for the documentation CMake project (if it's created independently). I'll make it an internal variable in the EnergyPlus CMake project.

nealkruis committed:
Remove test message.

nealkruis committed:
Make DOCS_TESTING an internal cache variable in the EnergyPlus CMake project.

Myoldmopar commented:
I don't see how this could've broken that one failing Windows file.  I'm feeling OK with just letting this merge as-is.  I've got a [branch of Decent CI](https://github.com/Myoldmopar/decent_ci/pull/42) that moves the "test errors" into "build errors" which is much more appropriate here. I won't worry about merging that immediately but I'll try some time soon.
